third_party/pcsc-lite/naclport/server_clients_management/src/permissions_checking/user-prompting-checker-unittests.js

```diff
@@ -0,0 +1,288 @@
+goog.require('GoogleSmartCard.PopupWindow.Server');
+goog.require('goog.Promise');
+goog.require('goog.json');
+goog.require('goog.testing');
+goog.require('goog.testing.MockControl');
+goog.require('goog.testing.jsunit');
+goog.require('goog.testing.mockmatchers');
+
+goog.setTestOnly();
+
+goog.scope(function() {
+
+/** @const */
+var GSC = GoogleSmartCard;
+
+/** @const */
+var PermissionsChecking =
+    GSC.PcscLiteServerClientsManagement.PermissionsChecking;
+/** @const */
+var KnownApp = PermissionsChecking.KnownApp;
+/** @const */
+var KnownAppsRegistry = PermissionsChecking.KnownAppsRegistry;
+/** @const */
+var UserPromptingChecker = PermissionsChecking.UserPromptingChecker;
+/** @const */
+var ignoreArgument = goog.testing.mockmatchers.ignoreArgument;
+
+/** @const */
+var FAKE_APP_1_ID = 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa';
+/** @const */
+var FAKE_APP_1_NAME = 'App Name 1';
+/** @const */
+var FAKE_APP_1_KNOWN_APP = new KnownApp(FAKE_APP_1_ID, FAKE_APP_1_NAME);
+/** @const */
+var FAKE_APP_2_ID = 'bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb';
+/** @const */
+var STORAGE_KEY = 'pcsc_lite_clients_user_selections';
+
+/**
+ * @enum {number}
+ */
+var MockedDialogBehavior = {
+  NOT_RUN: 0,
+  USER_APPROVES: 1,
+  USER_DENIES: 2,
+  USER_CANCELS: 3,
+};
+
+/**
+ * Sets up a mock instance of KnownAppsRegistry.
+ *
+ * This mock instance is only aware of |FAKE_APP_1_ID|.
+ * @param {!goog.testing.MockControl} mockControl
+ */
+function setUpKnownAppsRegistryMock(mockControl) {
+  var mockInstance = {
+    getById: function(id) {
+      if (id == FAKE_APP_1_ID)
+        return goog.Promise.resolve(FAKE_APP_1_KNOWN_APP);
+      return goog.Promise.reject();
+    }
+  };
+  /** @type {?} */
+  var mockedConstructor = mockControl.createConstructorMock(
+      PermissionsChecking, 'KnownAppsRegistry');
+  mockedConstructor().$returns(mockInstance);
+}
+
+/**
+ * Sets up mocks for get() and set() methods of the chrome.storage.local
+ * Extensions API.
+ * @param {!goog.testing.MockControl} mockControl
+ * @param {!goog.testing.PropertyReplacer} propertyReplacer
+ * @param {!Object} fakeInitialData Fake data to be returned from get().
+ * @param {Object} expectedDataToBeWritten If non-null, then the expected data
+ * to be passed into set(); if null, then set() isn't expected to be called.
+ */
+function setUpChromeStorageMock(
+    mockControl, propertyReplacer, fakeInitialData, expectedDataToBeWritten) {
+  propertyReplacer.set(
+      chrome, 'storage',
+      {local: {
+         get: mockControl.createFunctionMock('chrome.storage.local.get'),
+         set: mockControl.createFunctionMock('chrome.storage.local.set')}});
+
+  /** @type {?} */ chrome.storage.local.get;
+  chrome.storage.local.get(STORAGE_KEY, ignoreArgument).$does(
+      function(key, callback) {
+        callback(fakeInitialData);
+      });
+
+  /** @type {?} */ chrome.storage.local.set;
+  if (!goog.isNull(expectedDataToBeWritten))
+    chrome.storage.local.set(expectedDataToBeWritten);
```

Comment:
Does this already set up an expectation that .set will be called with the passed parameter ( and not called with anything else ), or do we need to call something like $does / $return / .. on it?

Reply:
Nope, the expectation is already set up. For instance, the tests fail if I repeat this line.